File: leetcode/46-permutations.js
// https://leetcode.com/problems/permutations/ (top-interview-questions)


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  
  const backtracking = (idx = 0) => {
    if (idx === nums.length - 1) {
      return [ [nums[idx]] ];
    }

    const subPermutations = backtracking(idx+1);
    
    const permutations = [];

    for (const subPermutation of subPermutations) {
      for (let i = 0; i <= subPermutation.length; i++) {
        permutations.push([...subPermutation.slice(0, i), nums[idx], ...subPermutation.slice(i)]);
      }
    }
    
    return permutations;
  }

  return backtracking();

};




/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute02 = function(nums) {
  if (nums.length === 1) return [[nums[0]]];
  
  const permutations = [];
  const permutation = permute(nums.slice(1));

  for (const perm of permutation) {
    for (let i = 0; i <= perm.length; i++) {
      permutations.push([...perm.slice(0, i), nums[0], ...perm.slice(i)]);
    }
  }

  return permutations;
};



const result = permute02([1, 2, 3]);

console.log(result);

File: leetcode/46-permutations.py
# https://leetcode.com/problems/permutations/ (top-interview-questions)

from typing import List


class Solution:

    def permute(self, nums: List[int]) -> List[List[int]]:
        def backtracking(i):
            if i == len(nums) - 1:
                return [[nums[i]]]

            permutations = list()
            subPermutations = backtracking(i+1)
            for subPermutation in subPermutations:
                for j in range(len(subPermutation)+1):
                    pCopy = subPermutation.copy()
                    pCopy.insert(j, nums[i])
                    permutations.append(pCopy)
            
            return permutations
        
        return backtracking(0)


    def permute01(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1:
            return [nums]

        permutations = list()
        chosen = [False] * len(nums)
        permutation = list()

        def backtracking():
            if len(permutation) == len(nums):
                permutations.append(permutation.copy())
                return
            
            for i in range(len(nums)):
                if(chosen[i]):
                    continue
                permutation.append(nums[i])
                chosen[i] = True
                backtracking()
                chosen[i] = False
                permutation.pop()
        
        backtracking()
        return permutations
    


    def permute02(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1: return [nums[:]]
        permutations = []
        
        for _ in range(len(nums)):
            num = nums.pop(0)
            sub_permutations = self.permute(nums)
            for sub_permutation in sub_permutations:
                sub_permutation.append(num)
            permutations.extend(sub_permutations)
            nums.append(num)
        
        return permutations


